includes/wc-update-functions.php

```diff
@@ -1773,3 +1773,123 @@ function wc_update_340_last_active() {
+		);
+
+		$wpdb->query( 'SET sql_safe_updates=1' );
+
+		$wpdb->query(
+			$wpdb->prepare(
+				"UPDATE wp_posts JOIN customers_map ON wp_posts.ID = customers_map.post_id SET wp_posts.post_author = customers_map.customer_id WHERE post_type IN ({$post_types_placeholders})",  // phpcs:ignore WordPress.WP.PreparedSQL.NotPrepared
+				$post_types
+			)
+		);
+
+		$wpdb->update( $wpdb->posts, array( 'post_author' => 0 ), array( 'post_type' => 'shop_order_refund' ) );
+	} else {
+		// If running the update from the wp-admin, copy data in batches being careful not to use more memory than allowed.
+		$admin_orders_sql = '';
+		$admin_orders     = get_transient( 'wc_update_350_admin_orders' );
+
+		if ( false === $admin_orders ) {
+			// Get the list of orders that we don't want to change as they belong to user ID 1.
+			$admin_orders = $wpdb->get_col(
+				$wpdb->prepare(
+					"SELECT ID FROM wp_posts p
+					INNER JOIN wp_postmeta pm ON p.ID = pm.post_id
+					WHERE post_type IN ({$post_types_placeholders}) AND meta_key = '_customer_user' AND meta_value = 1", // phpcs:ignore WordPress.WP.PreparedSQL.NotPrepared
+					$post_types
+				)
+			);
+
+			// Cache the list of orders placed by the user ID 1 as to large stores this query can be slow.
+			set_transient( 'wc_update_350_admin_orders', $admin_orders, DAY_IN_SECONDS );
+		}
+
+		if ( ! empty( $admin_orders ) ) {
+			$admin_orders_sql .= ' AND ID NOT IN (' . implode( ', ', $admin_orders ) . ') ';
+		}
+
+		// Query to get a batch of orders IDs to change.
+		$query = $wpdb->prepare(
+			// phpcs:disable WordPress.WP.PreparedSQL.NotPrepared
+			"SELECT ID FROM {$wpdb->posts}
+	        WHERE post_author = 1 AND post_type IN ({$post_types_placeholders}) $admin_orders_sql
+			LIMIT 1000",
+			$post_types
+			// phpcs:enable
+		);
+
+		while ( true ) {
+			// phpcs:ignore WordPress.WP.PreparedSQL.NotPrepared
+			$orders_to_update = $wpdb->get_col( $query );
+
+			// Exit loop if no more orders to update.
+			if ( ! $orders_to_update ) {
+				break;
+			}
+
+			$orders_meta_data = $wpdb->get_results(
+				// phpcs:ignore WordPress.WP.PreparedSQL.NotPrepared
+				"SELECT post_id, meta_value as customer_id FROM {$wpdb->postmeta} WHERE meta_key = '_customer_user' AND post_id IN (" . implode( ', ', $orders_to_update ) . ')'
+			);
+
+			// Exit loop if no _customer_user metas exist for the list of orders to update.
+			if ( ! $orders_meta_data ) {
+				break;
+			}
+
+			// Update post_author for a batch of orders.
+			foreach ( $orders_meta_data as $order_meta ) {
+				// Stop update execution and re-enqueue it if near memory limit.
+				if ( $updater instanceof WC_Background_Updater && $updater->is_memory_exceeded() ) {
+					return -1;
+				}
+
+				$wpdb->update( $wpdb->posts, array( 'post_author' => $order_meta->customer_id ), array( 'ID' => $order_meta->post_id ) );
+			}
+		}
+
+		// Set post_author to 0 instead of 1 on all shop_order_refunds.
+		while ( true ) {
```

Comment:
This could be split into a separate callback - then background processing will take care of queuing it after orders are done.

Reply:
Could you elaborate on why you are suggesting this? It seems to me that it is better to keep all database changes needed for this PR in a single function.

Reply:
@rodrigoprimo You have several while loops in sequence. And you're handling memory usage manually.

If `wc_update_350_order_customer_id` was split into multiple callbacks, e.g. `wc_update_350_order_customer_id_refunds` etc, it will call them in sequence and manage memory for you.

Does that make sense?

Reply:
Even if I split into two callbacks, I would still want to manage memory usage manually inside each while loop so I don't see any advantage to split.